# Lab 2. Search
# Task 2.2 8-Puzzy Problem
## Problem Descriptions
In this problem, we have a 3x3 grid filled with numbered tiles from 1 to 8 and a blank space. Tiles are jumbled up at the beginning, and we need to arrange all tiles until they are numerically in order with a blank space. The program aims to find a solution with the least movement to arrange tiles into the target order. Firstly, the problem is defined with state, action, goal test, and path cost.

1. State: Position tiles in 3x3 arrays with numbers 1-8 and 'E' for the blank space. These positions are represented by one string with dashes, and the sequence of string elements is corresponding to the position of tiles in the 3x3 array.

2. Action: Action is represented by movement of the blank tile up, down, left, and right. However, boundary conditions must be checked to decide which 4 moves are valid for a specific state.

3. Goal test: The goal test checks whether the current state is 'E-1-2-3-4-5-6-7-8'. If yes, the solution is found and the search is ended.

4. Path cost: Path cost is the number of moves used to arrange jumbled tiles to target order. Therefore, 1 cost represents 1 move.



## Implementation and Results

In [ ]:
!pip install simpleai
from simpleai.search import SearchProblem, astar, greedy, breadth_first, depth_first

import time


In [ ]:
# Class returning no.of misplaced tiles to solve the puzzle
class NoOfMisplacedTilesPuzzleSolver(SearchProblem):

      # Action method to get the list of the possible numbers that can be moved in to the empty space
      def actions(self, cur_state):
          rows = string_to_list(cur_state)
          row_empty, col_empty = get_location(rows, 'E')

          actions = []
          if row_empty > 0: actions.append(rows[row_empty - 1][col_empty])
          if row_empty < 2: actions.append(rows[row_empty + 1][col_empty])
          if col_empty > 0: actions.append(rows[row_empty][col_empty - 1])
          if col_empty < 2: actions.append(rows[row_empty][col_empty + 1])
          return actions

      # Return the resulting state after moving a piece to the empty space
      def result(self, state, action):
          rows = string_to_list(state)
          row_empty, col_empty = get_location(rows, 'E')
          row_new, col_new = get_location(rows, action)
          rows[row_empty][col_empty], rows[row_new][col_new] = rows[row_new][col_new], rows[row_empty][col_empty]
          return list_to_string(rows)

      # Returns true if a state is the goal state
      def is_goal(self, state):
        return state == GOAL

      # Returns the number of misplaced tiles
      def heuristic(self, state):
          rows = string_to_list(state)
          goal = string_to_list(GOAL)
          distance = sum([rows[i] != goal[i] for i in range(len(rows))]) - 1
          return distance


# Class returns estimate of Manhattan distance to solve the puzzle
class DistancePuzzleSolver(SearchProblem):
      # Action method to get the list of the possible numbers that can be moved in to the empty space
      def actions(self, cur_state):
          rows = string_to_list(cur_state)
          row_empty, col_empty = get_location(rows, 'E')

          actions = []
          if row_empty > 0: actions.append(rows[row_empty - 1][col_empty])
          if row_empty < 2: actions.append(rows[row_empty + 1][col_empty])
          if col_empty > 0: actions.append(rows[row_empty][col_empty - 1])
          if col_empty < 2: actions.append(rows[row_empty][col_empty + 1])
          return actions

      # Return the resulting state after moving a piece to the empty space
      def result(self, state, action):
          rows = string_to_list(state)
          row_empty, col_empty = get_location(rows, 'E')
          row_new, col_new = get_location(rows, action)
          rows[row_empty][col_empty], rows[row_new][col_new] = rows[row_new][col_new], rows[row_empty][col_empty]
          return list_to_string(rows)

      # Returns true if a state is the goal state
      def is_goal(self, state):
        return state == GOAL

      # # Returns an estimate of the distance from a state to the goal using the manhattan distance
      def heuristic(self, state):
          rows = string_to_list(state)
          distance = 0
          for number in '12345678E':
               row_new, col_new = get_location(rows, number)
               row_new_goal, col_new_goal = goal_positions[number]
               distance += abs(row_new - row_new_goal) + abs(col_new - col_new_goal)
          return distance

In [ ]:
# Convert list to string
def list_to_string(input_list):
    return '\n'.join(['-'.join(x) for x in input_list])

# Convert string to list
def string_to_list(input_string):
    return [x.split('-') for x in input_string.split('\n')]

# Find the 2D location of the input element
def get_location(rows, input_element):
    for i, row in enumerate(rows):
        for j, item in enumerate(row):
            if item == input_element:
                return i, j

# Final result that we want to achieve
GOAL = '''\
E-1-2
3-4-5
6-7-8'''

# Create a cache for the goal position of each piece
goal_positions = {}
rows_goal = string_to_list(GOAL)
for number in '12345678E':
    goal_positions[number] = get_location(rows_goal, number)

# Print the results
def print_result_path(result):
  for i, (action, state) in enumerate(result.path()):
      print()
      if action == None:
        print('Initial configuration')
      else:
        print('Step %s: After moving %s into the empty space' %(i, action))
  print(state)

In [ ]:
# Simple case Starting point - solution depth = 4 (no of misplaced tiles)
INITIAL = '''\
1-4-2
3-5-8
6-7-E'''
%time astarresult=astar(NoOfMisplacedTilesPuzzleSolver(INITIAL))
print_result_path(astarresult)
print('Goal of simple case achieved with A* search using heuristic with the number of misplaced tiles!\n\n')


%time greedyresult=greedy(NoOfMisplacedTilesPuzzleSolver(INITIAL))
print_result_path(greedyresult)
print('Goal of simple case achieved with greedy search using heuristic with the number of misplaced tiles!\n\n')


%time breadth_firstresult=breadth_first(NoOfMisplacedTilesPuzzleSolver(INITIAL))
print_result_path(breadth_firstresult)
print('Goal of simple case achieved with breadth-first search using heuristic with the number of misplaced tiles!\n\n')

#%time depth_firstresult = depth_first(NoOfMisplacedTilesPuzzleSolver(INITIAL))
#print_result_path(depth_firstresult)
#print('Goal of simple case achieved with breadth-first search using heuristic with the number of misplaced tiles!')

#Depth first search is commented because it gaves runtime error when solving the puzzle problem.


CPU times: user 490 µs, sys: 0 ns, total: 490 µs
Wall time: 508 µs

Initial configuration

Step 1: After moving 8 into the empty space

Step 2: After moving 5 into the empty space

Step 3: After moving 4 into the empty space

Step 4: After moving 1 into the empty space
E-1-2
3-4-5
6-7-8
Goal of simple case achieved with A* search using heuristic with the number of misplaced tiles!


CPU times: user 406 µs, sys: 0 ns, total: 406 µs
Wall time: 416 µs

Initial configuration

Step 1: After moving 8 into the empty space

Step 2: After moving 5 into the empty space

Step 3: After moving 4 into the empty space

Step 4: After moving 1 into the empty space
E-1-2
3-4-5
6-7-8
Goal of simple case achieved with greedy search using heuristic with the number of misplaced tiles!


CPU times: user 668 µs, sys: 803 µs, total: 1.47 ms
Wall time: 1.62 ms

Initial configuration

Step 1: After moving 8 into the empty space

Step 2: After moving 5 into the empty space

Step 3: After moving 4 into the empty s

In [ ]:
# Simple case Starting point - solution depth = 4 ( Manhattan distance)
INITIAL = '''\
1-4-2
3-5-8
6-7-E'''

%time astarresult = astar(DistancePuzzleSolver(INITIAL))
print_result_path(astarresult)
print('Goal of simple case achieved with A* search using heuristic with the sum of distances of tiles from their target position!\n\n')

%time greedyresult = greedy(DistancePuzzleSolver(INITIAL))
print_result_path(greedyresult)
print('Goal of simple case achieved with greedy search using heuristic with the sum of distances of tiles from their target position!\n\n')

%time breadth_firstresult = breadth_first(DistancePuzzleSolver(INITIAL))
print_result_path(breadth_firstresult)
print('Goal of simple case achieved with breadth-first search using heuristic with the sum of distances of tiles from their target position\n\n!')

#%time depth_firstresult = depth_first(DistancePuzzleSolver(INITIAL))
#print_result_path(depth_firstresult)
#print('Goal of simple case achieved with breadth-first search using heuristic  with the sum of distances of tiles from their target position\n\n!')

#Depth first search is commented because it gaves runtime error when solving the puzzle problem.

CPU times: user 600 µs, sys: 0 ns, total: 600 µs
Wall time: 605 µs

Initial configuration

Step 1: After moving 8 into the empty space

Step 2: After moving 5 into the empty space

Step 3: After moving 4 into the empty space

Step 4: After moving 1 into the empty space
E-1-2
3-4-5
6-7-8
Goal of simple case achieved with A* search using heuristic with the sum of distances of tiles from their target position!


CPU times: user 633 µs, sys: 0 ns, total: 633 µs
Wall time: 649 µs

Initial configuration

Step 1: After moving 8 into the empty space

Step 2: After moving 5 into the empty space

Step 3: After moving 4 into the empty space

Step 4: After moving 1 into the empty space
E-1-2
3-4-5
6-7-8
Goal of simple case achieved with greedy search using heuristic with the sum of distances of tiles from their target position!


CPU times: user 1.29 ms, sys: 0 ns, total: 1.29 ms
Wall time: 1.32 ms

Initial configuration

Step 1: After moving 8 into the empty space

Step 2: After moving 5 into the

In [ ]:
# Difficult caseStarting point - solution depth = 26 by estimating sum of Manhattan distance
DIFF_INITIAL = '''\
7-2-4
5-E-6
8-3-1'''

%time astarresult = astar(DistancePuzzleSolver(DIFF_INITIAL))
print_result_path(astarresult)
print('Goal of difficult case achieved with A* search using heuritic with the sum of distances of tiles from their target position!')

#%time greedyresult = greedy(DistancePuzzleSolver(DIFF_INITIAL))
#print_result_path(greedyresult)
#print('Goal achieved with greedy search!')

##greedy search is commented because it gives runtime error




CPU times: user 21.9 s, sys: 368 ms, total: 22.3 s
Wall time: 22.9 s

Initial configuration

Step 1: After moving 5 into the empty space

Step 2: After moving 7 into the empty space

Step 3: After moving 2 into the empty space

Step 4: After moving 5 into the empty space

Step 5: After moving 6 into the empty space

Step 6: After moving 1 into the empty space

Step 7: After moving 3 into the empty space

Step 8: After moving 8 into the empty space

Step 9: After moving 7 into the empty space

Step 10: After moving 6 into the empty space

Step 11: After moving 1 into the empty space

Step 12: After moving 3 into the empty space

Step 13: After moving 8 into the empty space

Step 14: After moving 7 into the empty space

Step 15: After moving 6 into the empty space

Step 16: After moving 1 into the empty space

Step 17: After moving 3 into the empty space

Step 18: After moving 4 into the empty space

Step 19: After moving 5 into the empty space

Step 20: After moving 2 into the empty spa

In [ ]:
# Difficult caseStarting point - solution depth = 26 (No of misplaced tiles)
DIFF_INITIAL = '''\
7-2-4
5-E-6
8-3-1'''

%time astarresult = astar(NoOfMisplacedTilesPuzzleSolver(DIFF_INITIAL))
print_result_path(astarresult)
print('Goal achieved with A* search!')

%time greedyresult = greedy(NoOfMisplacedTilesPuzzleSolver(DIFF_INITIAL))
print_result_path(greedyresult)
print('Goal achieved with greedy search!')

#Both A* and greedy best first search gaves error runtime when solving the difficult case.

## Discussions
###Simple case
A* search, greedy best-first search, breadth-first search and depth-first search have been implemented to solve the simple case '1-4-2-3-5-8-6-7-E' of 4 solution depth. All search algorithms generate a solution with 4 moves except depth-first search, which generates a run time error using both heuristic functions.

Looking at the execution time of various search algorithms using heuristic with the number of misplaced tiles, A* search was completed in 490 µs, greedy best-first search in 406 µs, and breadth-first search in 1.47 ms. However, a run time error is occurred when a depth-first search is executed. As a result, greedy best-first search is the fastest search algorithm to solve the simple case of 4 solution depth using heuristic with the number of misplaced tiles, while breadth-first search is the slowest, and depth-first search is impractical.

By observing the execution time of each search algorithm using heuristic with Manhattan distance, A* search was executed in 600 µs, greedy best-first search in 633 µs, and breadth-first search in 1.29 ms. However, a run time error occurred once again when the depth-first search was implemented. Therefore, A* search is the fastest search algorithm to solve the simple case of 4 solution depth using this heuristic, while breadth-first search is the slowest, and depth-first search is impractical.



###Difficult case
A* search and greedy best-first search have been implemented to solve the difficult case '7-2-4-5-E-6-8-3-1' using both heuristic functions. Both A* search and greedy best-first search give a run time error using a heuristic with the number of misplaced tiles. When using the heuristic with Manhattan distance, A* search generates a solution of 26 moves in 22.3s, while greedy best-first search gives a run time error once again.

The potential reason for run time error might be due to either a poor heuristic function being implemented, rendering it challenging to find the solution or an excessive number of generated nodes during the search exceeding control limits, consuming lots of memory and causing crashed at the end. Another potential reason is that search algorithm get trapped in an infinity loop during puzzle solving.

In summary, only A star search employing with heuristic based on the Manhattan distance can generate a solution for a difficult case of 26 solution depth. Further investigation is required to address run-time error.
